overlord/snapstate/check_snap.go

```diff
@@ -478,6 +479,22 @@ func checkGadgetOrKernel(st *state.State, snapInfo, curInfo *snap.Info, _ snap.C
 		return nil
 	}
 
+	// do basic constraints check on the gadget
+	model := deviceCtx.Model()
+	if typ == snap.TypeGadget {
+		validateGadgetModelConstraints := func(model *asserts.Model, snapf snap.Container) error {
+			constraints := &gadget.ModelConstraints{
+				Classic:    model.Classic(),
+				SystemSeed: model.Grade() != asserts.ModelGradeUnset,
+			}
+			_, err = gadget.ReadInfoFromSnapFile(snapf, constraints)
```

Comment:
this definitely shows that turning ModelConstraints into an interface subsetting asserts.Model is probably a cleaner path forward

Reply:
I will explore this now.